### PR TITLE
add new zabbix module - zabbix_discovery_rule

### PR DESCRIPTION
**Pull request by tcraxs**
##### SUMMARY

Add new module for create, delete and update Zabbix discovery rules.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME

zabbix_discovery_rule

##### ADDITIONAL INFORMATION

- Include all available API options in this module. Reference can be found here: https://www.zabbix.com/documentation/current/manual/api/reference/drule
- Tests for the new module are included.

Below is the documentation of the module:

```paste below
module: zabbix_discovery_rule
short_description: Create/delete/update Zabbix discovery rules
description:
   - Create discovery rule.
   - Delete existing discovery rule.
   - Update existing discovery rule with new options.
author:
    - "Tobias Birkefeld (@tcraxs)"
requirements:
    - zabbix-api
options:
    state:
        description:
            - Create or delete user group.
        required: false
        type: str
        default: "present"
        choices: [ "present", "absent" ]
    name:
        description:
            - Name of the discovery rule.
        required: true
        type: str
    iprange:
        description:
            - One or several IP ranges to check separated by commas.
        required: true
        type: list
        elements: str
    dchecks:
        description:
            - List of dictionaries of discovery check objects.
            - For more information, review discovery check object documentation at
              U(https://www.zabbix.com/documentation/current/manual/api/reference/dcheck/object)
        required: true
        suboptions:
            type:
                description:
                    - Type of check.
                    - "Possible values:"
                    - "0 - SSH"
                    - "1 - LDAP"
                    - "2 - SMTP"
                    - "3 - FTP"
                    - "4 - HTTP"
                    - "5 - POP"
                    - "6 - NNTP"
                    - "7 - IMAP"
                    - "8 - TCP"
                    - "9 - Zabbix agent"
                    - "10 - SNMPv1 agent"
                    - "11 - SNMPv2 agent"
                    - "12 - ICMP ping"
                    - "13 - SNMPv3 agent"
                    - "14 - HTTPS"
                    - "15 - Telnet"
                type: int
                choices: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
                required: true
            ports:
                description:
                    - One or several port ranges to check separated by commas. Used for all checks except for ICMP.
                type: str
            key_:
                description:
                    - "The value of this property differs depending on the type of the check:"
                    - "- key to query for Zabbix agent checks"
                    - "- SNMP OID for SNMPv1, SNMPv2 and SNMPv3 checks"
                type: str
            snmp_community:
                description:
                    - SNMP community.
                    - Required for SNMPv1 and SNMPv2 agent checks.
                type: str
            snmpv3_authpassphrase:
                description:
                    - Authentication passphrase used for SNMPv3 agent checks with security level set to authNoPriv or authPriv.
                type: str
            snmpv3_authprotocol:
                description:
                    - Authentication protocol used for SNMPv3 agent checks with security level set to authNoPriv or authPriv.
                    - "Possible values:"
                    - "0 - (default) MD5"
                    - "1 - SHA"
                type: int
                choices: [0, 1]
            snmpv3_contextname:
                description:
                    - SNMPv3 context name. Used only by SNMPv3 checks.
                type: str
            snmpv3_privpassphrase:
                description:
                    - Privacy passphrase used for SNMPv3 agent checks with security level set to authPriv.
                type: str
            snmpv3_privprotocol:
                description:
                    - Privacy protocol used for SNMPv3 agent checks with security level set to authPriv.
                    - "Possible values:"
                    - "0 - (default) DES"
                    - "1 - AES"
                type: int
                choices: [0, 1]
            snmpv3_securitylevel:
                description:
                    - Security level used for SNMPv3 agent checks.
                    - "Possible values:"
                    - "0 - noAuthNoPriv"
                    - "1 - authNoPriv"
                    - "2 - authPriv"
                type: int
                choices: [0, 1, 2]
            snmpv3_securityname:
                description:
                    - Security name used for SNMPv3 agent checks.
                type: str
            uniq:
                description:
                    - Whether to use this check as a device uniqueness criteria.
                    - Only a single unique check can be configured for a discovery rule.
                    - Used for Zabbix agent, SNMPv1, SNMPv2 and SNMPv3 agent checks.
                    - "Possible values:"
                    - "0 - (default) do not use this check as a uniqueness criteria"
                    - "1 - use this check as a uniqueness criteria"
                type: int
                default: 0
                choices: [0, 1]
            host_source:
                description:
                    - Source for host name.
                    - "Possible values:"
                    - "1 - (default) DNS"
                    - "2 - IP"
                    - "3 - discovery value of this check"
                type: int
                default: 1
                choices: [1, 2, 3]
            name_source:
                description:
                    - Source for visible name.
                    - "Possible values:"
                    - "0 - (default) not specified"
                    - "1 - DNS"
                    - "2 - IP"
                    - "3 - discovery value of this check"
                type: int
                default: 0
                choices: [0, 1, 2, 3]
        type: list
        aliases: [ "dcheck" ]
    delay:
        description:
            - Execution interval of the discovery rule.
        required: false
        type: str
        default: "1h"
    proxy:
        description:
            - Name of the proxy used for discovery.
        required: false
        type: str
    status:
        description:
            - Whether the discovery rule is enabled.
            - "Possible values:"
            - "0 - (default) enabled"
            - "1 - disabled"
        required: false
        type: int
        default: 0
        choices: [0, 1]
```

### CODE REVIEW

**Comment by tcraxs** [comment]
@D3DeFi Thanks for review and your comments. I will adjust these parts. With Zabbix 3.0 I have some more problems beside the `delay` which has to be an integer. I will show it to you later.

**Comment by D3DeFi** [comment]
> 
> 
> @D3DeFi Thanks for review and your comments. I will adjust these parts. With Zabbix 3.0 I have some more problems beside the `delay` which has to be an integer. I will show it to you later.

Np, feel free to stop by our [gitter chat](https://gitter.im/community-zabbix/community) if you have more questions or need advice, I will be there later today in a 3 hours or so

File: plugins/modules/zabbix_discovery_rule.py [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+        description:
+            - List of dictionaries of discovery check objects.
+            - For more information, review discovery check object documentation at
+              U(https://www.zabbix.com/documentation/current/manual/api/reference/dcheck/object)
+        suboptions:
+            type:
+                description:
+                    - Type of check.
+                    - "Possible values:"
+                    - "0 - SSH"
```
[the remark this answers is not in the document]
Comment (sky-joker) [reply plugins/modules/zabbix_discovery_rule.py:55]: I think it's easier for people to understand a string than numbers :)  
It is recommended to change it to something like `zabbix_action`.

File: plugins/modules/zabbix_discovery_rule.py [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+# -*- coding: utf-8 -*-
+#
+# (c) 2013-2014, Epic Games, Inc.
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
```
Comment (sky-joker) [inline plugins/modules/zabbix_discovery_rule.py:11]: Can you please remove `ANSIBLE_METADATA`?  
`ANSIBLE_METADATA` is no longer necessary.

see:
https://github.com/ansible-collections/overview/issues/57
https://github.com/ansible-collections/community.zabbix/pull/48

File: plugins/modules/zabbix_discovery_rule.py [hunk 3]
```diff
@@ -0,0 +1,700 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# (c) 2013-2014, Epic Games, Inc.
```
Comment (D3DeFi) [inline plugins/modules/zabbix_discovery_rule.py:4]:
```suggestion
```

**Comment by tcraxs** [comment]
> We will probably later strip the module a bit due to initiative in #104, but don't worry about it now.
I can do this later, when this PR is merged.

**Comment by sky-joker** [comment]
@D3DeFi 

It would be nice to add a RETURN block to this module, do you want to merge?
I thought it would be useful for users of the module if you return the value you set in the RETURN block.

**Comment by tcraxs** [comment]
> @D3DeFi
> 
> It would be nice to add the RETURN block to this module, do you want to merge?
> I thought it would be useful for users of the module if you return the value you set in the RETURN block.
> 
> https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#return-block

Thanks for the hint. Yes, I think a RETURN block could be nice. I will add a RETURN-block with additional infos what changed or so.
Is it okay for you to add this RETURN-block after with the refactoring in terms of initiative #104 ?

**Comment by sky-joker** [comment]
> Thanks for the hint. Yes, I think a RETURN block could be nice. I will add a RETURN-block with additional infos what changed or so.
Is it okay for you to add this RETURN-block after with the refactoring in terms of initiative #104 ?

Thank you @tcraxs.
Yes, please do!

**Pull request merged**